src/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs

```diff
@@ -292,37 +223,68 @@ private void FlushMessageQueue()
                 // We need to take lock before DeQueue.
                 if (_messageQueue.Count > 0)
                 {
+                    string line = null;
+                    // We need this flag to ensure that line was taken from the queue.
+                    // We couldn't just check line on null because ReadBuffer enqueue null in case of EOF.
+                    bool isDequeue = false;
                     lock (_messageQueue)
                     {
                         if (_messageQueue.Count > 0)
                         {
-                            string s = (string)_messageQueue.Dequeue();
-                            // skip if the read is the read is cancelled
-                            // this might happen inside UserCallBack
-                            // However, continue to drain the queue
-                            if (!_cancelOperation)
-                            {
-                                _userCallBack(s);
-                            }
+                            line = _messageQueue.Dequeue();
+                            isDequeue = true;
                         }
                     }
+
+                    // skip if the read is cancelled
+                    // this might happen inside UserCallBack
+                    // However, continue to drain the queue
+                    if (!_cts.Token.IsCancellationRequested && isDequeue)
+                    {
+                        return CallUserCallback(line, rethowInNewthread);
+                    }
                 }
                 else
                 {
                     break;
                 }
             }
+
+            return false;
+        }
+
+        // If user's callback throw we re-throw.
+        private bool CallUserCallback(string argument, bool rethowInNewthread)
+        {
+            try
+            {
+                _userCallBack(argument);
+            }
+            catch (Exception e)
+            {
+                if (rethowInNewthread)
+                {
+                    ExceptionDispatchInfo edi = ExceptionDispatchInfo.Capture(e);
+                    Task.Run(() => edi.Throw());
```

Comment:
This will need to be ThreadPool.QueueUserWorkItem rather than Task.Run, as Task.Run will store the exception into the returned task (which is ignored).